### ISSUE TITLE
[MJAVADOC-775] Option 'taglets/taglet/tagletpath' ignored when pointing to a JAR

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-javadoc-plugin#1153]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** opened **[MJAVADOC-775](https://issues.apache.org/jira/browse/MJAVADOC-775?redirect=false)** and commented

The [`tagletpath` property](https://maven.apache.org/plugins/maven-javadoc-plugin/apidocs/org/apache/maven/plugins/javadoc/options/Taglet.html#getTagletpath()) used in a [`taglets`/`taglet` element](https://maven.apache.org/plugins/maven-javadoc-plugin/javadoc-mojo.html#taglets) seems to be completely ignored, despite being documented. It looks as if this never worked before. It is also utterly untested. There is an old bug MJAVADOC-231, but whatever that was meant to fix, it seems to be unrelated to this one. If I have a local JAR unavailable on Maven Central sitting in a libraries folder, I cannot use it in any other way than to contrive to put a copy into my local Maven repository and use `tagletArtifact` instead of `taglets`/`taglet`/`tagletpath`.

I noticed this, when trying to help someone asking a question on the [Maven users mailing list](https://lists.apache.org/thread/zv25z3hx9gbmbt6wwhby6cy0t5lr542l). In his [sample repository](https://github.com/jkesselm/xalan-java-mvn), an effort to convert the Xalan Ant Build to Maven, the commit to check out is is [6daf2890](https://github.com/jkesselm/xalan-java-mvn/tree/6daf2890bc03b13cd741b963d33897d61dcf4e5f). Just run `mvn -DskipTests=true site`.

---

**Affects:** 3.6.0, 3.6.3

**Issue Links:**
- [MJAVADOC-783](https://issues.apache.org/jira/browse/MJAVADOC-783) Invalid path when using TagletArtifact and TagletPath
 (_**"duplicates"**_)

**Remote Links:**
- [GitHub Pull Request #255
](https://github.com/apache/maven-javadoc-plugin/pull/255)

0 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** commented

If that's so, we should probably just remove tagletpath from the docs

**Comment by jira-importer** [comment]
**[Rob Gordon](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER303448)** commented

I have raised MJAVADOC-783 regarding a bug in TagletPath. This may be related to this issue.

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo), I was actually hoping for the feature to be made working, because it would do something valuable if it worked. I am not sure it is a good idea to simply hide something in the docs that exists as a plugin option or to completely remove the whole thing, just because nobody feels like fixing it.

**Comment by jira-importer** [comment]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** commented

Duping against 783 since that includes a little more info for a potential fix.

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo), are you sure it is is good idea to close this one? My `ClassNotFoundException` is not the same as the `InvalidPathException` in the other issue. The descriptions also sound different. Just because there happen to be two issues about tagle paths, it does not automatically mean that they are identical.

I also do not understand what you are missing here. There is a link to a real-world reproducer with exact instructions which commit to check out. A link to previous conversation on a mailing list about this, where you can also find an error description, also exists. Usually, developers ask for reproducers and links to further details. Sometimes, when providing same, I feel like they have not been read.

**Comment by jira-importer** [comment]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** commented

Bug triage and maintenance is important. We have plenty of evidence now that taglet path simply doesn't work. Thanks. Now someone needs to produce a test and a fix. If it turns out this problem still isn't fixed after that, we can reopen, but for now let's not keep two issues open for what really smells like two manifestations of the same issue.

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

Understood.

**Comment by jira-importer** [comment]
**[Alexander Kriegisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kriegaex)** commented

> Bug triage and maintenance is important.

[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo), then why did you not triage, comparing the situations in the two issues? 😉 This is **not** a duplicate, I just checked, adding some debug statements to the javadoc plugin, and finally realised that
1. there are two situations in which `<tagletpath>my/path</tagletpath>` can appear in the plugin configuration - namely (top level) option `tagletpath` and `{}taglets{`}/`{}taglet{`}/`{}tagletpath{`}) - and
2. therefore, my issue description falsely linked to the former (mea culpa, I just fixed that), while, as you would have seen if you had inspected and run my reproducer, I am using the latter.

MJAVADOC-783 is about path concatenation when consolidating those two plus `{}tagletartifacts{`}, creating a single taglet path for the javadoc CLI option `-tagletpath` from all 3 sources. 783 is about an omitted path separator and exists independently of the problem described here.

I managed to separately reproduce and fix both issues idependently of each other.

The fix for MJAVADOC-783 (as suggested similarly by [robjg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=robjg)) is:

```none
--- a/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java
+++ b/src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java
@@ -2956,7 +2956,7 @@ public abstract class AbstractJavadocMojo extends AbstractMojo {
         path.append(StringUtils.join(pathParts.iterator(), File.pathSeparator));
 
         if (tagletpath != null && !tagletpath.isEmpty()) {
-            path.append(JavadocUtil.unifyPathSeparator(tagletpath));
+            path.append(path.length() > 0 ? File.pathSeparator : "").append(JavadocUtil.unifyPathSeparator(tagletpath));
         }
 
         return path.toString();
```

The fix for this issue MJAVADOC-775 is something similar to:

```none
--- a/src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
+++ b/src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
@@ -136,7 +136,8 @@ public class JavadocUtil {
 
             Path directory = projectBasedir.resolve(dir);
 
-            if (Files.isDirectory(directory)) {
+            if (Files.isDirectory(directory)
+                    || directory.getFileName().toString().endsWith(".jar")) {
                 pruned.add(directory.toAbsolutePath());
             }
         }
```

[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo), maybe the check for ".jar" at the end of the path name is too naive. Maybe we need to consider other packaging types, too, or maybe we should not remove non-directory entries at all and just verify that the given paths actually exist, not matter if they are directories or files. WDYT? And would you like one PR fixing both issues or one per issue? Sorry, I just had a small time box to debug this, hoping that actually a Maven Javadoc committer would do it. So, other than the external reproducer project linked to, I have not created any unit or integration tests for this. I could open two PR drafts and your could commit on top of them. Now that the root causes of both issues are clear, it should be easy to create regression tests for them.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Reopening to give this issue more visibilty, maybe it is a dup, but I rather use my cycles in addressing the issue than figuring out the dup fact.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [91369fabe400ea6c2d91a36d90885d36b447d3ec](https://gitbox.apache.org/repos/asf?p=maven-javadoc-plugin.git;a=commit;h=91369fabe400ea6c2d91a36d90885d36b447d3ec).
